membership-attribute-service/app/controllers/BehaviourController.scala

```diff
@@ -23,30 +20,63 @@ class BehaviourController extends Controller with LazyLogging {
   lazy val metrics = Metrics("BehaviourController")
 
   def capture() = BackendFromCookieAction.async { implicit request =>
-    awsAction(request, "add")
+    awsAction(request, "upsert")
   }
 
   def remove = BackendFromCookieAction.async { implicit request =>
     awsAction(request, "delete")
   }
 
+  def sendCartReminderEmail = backendAction.async { implicit request =>
+    val receivedBehaviour = behaviourFromBody(request.body.asJson)
+    for {
+      paidTier <- request.touchpoint.attrService.get(receivedBehaviour.userId).map(_.exists(_.isPaidTier))
+      user <- request.touchpoint.identityService.user(IdentityId(receivedBehaviour.userId))
+      emailAddress = (user \ "user" \ "primaryEmailAddress").as[String]
+      gnmMarketingPrefs = true // TODO - needs an Identity API PR to send statusFields.receiveGnmMarketing in above user <- ... call
+    } yield {
+        val msg = if (paidTier || !gnmMarketingPrefs) {
+          logger.info(s"### NOT sending email because paidTier: $paidTier gnmMarketingPrefs: $gnmMarketingPrefs")
+          request.touchpoint.behaviourService.delete(receivedBehaviour.userId)
+          logger.info(s"### deleted reminder record")
+          "user has paid or is not accepting emails"
+        } else {
+          logger.info(s"### sending email to $emailAddress (TESTING ONLY - NO EMAIL IS ACTUALLY GENERATED YET!)")
+          // TODO!
+          // compile and send the email here
+          logger.info(s"### updating behaviour record emailed: true")
+          request.touchpoint.behaviourService.set(receivedBehaviour.copy(emailed = Some(true)))
+          "email sent - reminder record updated"
+        }
+      Ok(msg)
+    }
+  }
+
   private def awsAction(request: BackendRequest[AnyContent], action: String) = {
-    val behaviour = behaviourFromBody(request.body.asJson)
+    val receivedBehaviour = behaviourFromBody(request.body.asJson)
     action match {
-      case "add" =>
-        val addResult = for {
-          addItemResult <- request.touchpoint.behaviourService.set(behaviour)
-        } yield addItemResult
-        addResult.map { r =>
-          logger.info(s"recorded ${behaviour.activity} on ${behaviour.lastObserved} for ${behaviour.userId} with note ${behaviour.note}")
-          Ok(Behaviour.asJson(behaviour))
+      case "upsert" =>
+        val updateResult = for {
+          current <- request.touchpoint.behaviourService.get(receivedBehaviour.userId)
+          upserted = current.map { bhv =>
+            bhv.copy(
+              activity = receivedBehaviour.activity.orElse(bhv.activity),
+              lastObserved = receivedBehaviour.lastObserved.orElse(bhv.lastObserved),
+              note = receivedBehaviour.note.orElse(bhv.note),
+              emailed = receivedBehaviour.emailed.orElse(bhv.emailed))
+          }.getOrElse(receivedBehaviour)
+          res <- request.touchpoint.behaviourService.set(upserted)
```

Comment:
This upserting code is kind of interesting to me because it looks like a place where maybe the [DynamoDB upsert functionality](http://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_UpdateItem.html) would be useful? Under the hood, we are using Scanamo here, and I think Scanamo has some kind of functionality around this: https://github.com/guardian/scanamo#updating